SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,435 @@
+  let storageKind: AccessStorageKind
+  
+  var description: String {
+    "\(storageKind) - \(baseAddress)"
+  }
+  
+  var referenceWithPathComponent: (ref: Value, comp: (SmallProjectionPath.FieldKind, Int))? {
+    switch storageKind {
+    case .Box, .Tail, .Class:
+      switch baseAddress {
+      case let rea as RefElementAddrInst:
+        return (rea.operand, (.classField, rea.fieldIndex))
+      case let pb as ProjectBoxInst:
+        return (pb.operand, (.classField, pb.fieldIndex))
+      case let rta as RefTailAddrInst:
+        return (rta.operand,(.tailElements, 0))
+      default:
+        fatalError("Inconsistent state")
+      }
+    default:
+      return nil
+    }
+  }
+  
+  var reference: Value? { referenceWithPathComponent?.ref }
+  
+  var isLet: Bool {
+    switch baseAddress {
+    case is RefElementAddrInst:
+      fatalError() // FIXME: add support for looking up types in libswift etc
+    default:
+      fatalError()
+    }
+  }
+  
+  var isLocal: Bool {
+    switch storageKind {
+    case .Box:
+      return baseAddress is AllocBoxInst
+    case .Stack:
+      return true
+    default:
+      return false
+    }
+  }
+  
+  /// Returns `true` if we can reliably compare for equality this `AccessBase`
+  /// with another `AccessBase`
+  var isUniquelyIdentified: Bool {
+    switch storageKind {
+    case .Box:
+      return baseAddress is AllocBoxInst
+    case .Stack, .Global:
+      return true
+    case .Argument:
+      return (baseAddress as! FunctionArgument).isExclusiveIndirectParameter
+    default:
+      return false
+    }
+  }
+  
+  func isDistinct(from other: AccessBase) -> Bool {
+    fatalError()
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+  
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+  
+  func isDistinct(from: AccessPath) -> Bool {
+    fatalError("TODO")
+  }
+}
+
+/// An `AccessRoot` is the reference (or a value for which the
+/// reference is a stored property) an address originates from.
+struct AccessRoot {
+  var root: Value
+  
+  /// Only valid paths are: `"<sequence of value projections>.<one reference projection>.<sequence of address projections>"`
+  var path: SmallProjectionPath
+}
+
+typealias AccessStorage = (maybeRoot: AccessRoot?, base: AccessBase)
+
+/// An `AccessScope` denotes a pair of `begin_access` `end_access`
+/// instructions.
+struct AccessScope {
+  // TODO: add support for begin_access_unpaired
+  var beginAccess: BeginAccessInst
+
+  var accessedAddress: Value { beginAccess.operand }
+
+  var accessKind: AccessKind { beginAccess.accessKind }
+  
+  var scopeEndingInstructions: EndAccessInstructions {
+    beginAccess.endAccesses
+  }
+}
+
+enum EnclosingScope {
```

Comment:
Why do we need this enum?
This can be represented as a simple AccessBase, either a Nested or a not-Nested.